### PR TITLE
[3.1] Add auto-repair for locked/corrupted git repos

### PR DESCRIPTION
**Pull request by phuongnq**
### Ticket reference or full description of what's in the PR
https://github.com/craftercms/craftercms/issues/5600

### CODE REVIEW

File: src/main/java/org/craftercms/deployer/impl/processors/git/GitPullProcessor.java [hunk 1]
```diff
@@ -120,12 +121,19 @@ protected void doPull(ProcessorExecution execution) throws DeployerException {
 
             execution.setStatusDetails(details);
         } catch (JGitInternalException e) {
-            Throwable cause = e.getCause();
-            if (cause instanceof CorruptObjectException || cause instanceof EOFException) {
+            if (isRepositoryLocked(e)) {
+                logger.warn("The local repository {} is locked, trying to unlock it", localRepoFolder);
+                try {
+                    GitUtils.unlock(localRepoFolder.getAbsolutePath());
+                    logger.info(".git/index.lock is deleted from local repository {}", localRepoFolder);
```
Comment (sumerjabri) [inline src/main/java/org/craftercms/deployer/impl/processors/git/GitPullProcessor.java]: Please use '{}' as a standard placeholder (I'm trying to have the whole system be consistent).

**Review by sumerjabri — CHANGES_REQUESTED** [review]
I think we need to clean things up on startup only, not on every pull. @jdrossl correct me if I'm wrong here.

**Comment by jdrossl** [comment]
I'm ok with checking on every pull, it integrates with how the system recovers from the corrupted index

**Review by jdrossl — APPROVED** [review]

**Comment by sumerjabri** [comment]
> I'm ok with checking on every pull, it integrates with how the system recovers from the corrupted index

What about an externally held lock?

@avasquez614 @russdanner how realistic is it to have an externally held lock? Should we do this on startup only or on every pull? (this is Deployer)

**Review by sumerjabri — CHANGES_REQUESTED** [review]
Please make it clean up on start up only.

**Comment by phuongnq** [comment]
Modified so that on startup, check for targets that need unlock.

**Review by jdrossl — APPROVED** [review]

**Review by sumerjabri — APPROVED** [review]

**Pull request merged**